typing/env.ml

```diff
@@ -3655,8 +3664,44 @@ let report_lookup_error_doc _loc env ppf = function
         "The ancestor variable %a@ \
          cannot be accessed from the definition of an instance variable"
        quoted_longident lid
-  | Illegal_reference_to_recursive_module ->
-     fprintf ppf "Illegal recursive module reference"
+  | Illegal_reference_to_recursive_module { container; unbound } ->
+      let container = Option.value ~default:"_" container in
+      let self_or_unbound =
+        if String.equal container unbound
+        then dprintf "itself"
+        else dprintf "the module type of %a" Style.inline_code unbound
+      in
+      fprintf ppf
+        "@[<hov>This module type is recursive.@ \
+         This use of the recursive module %a@ \
+         within the definition of the module %a@ \
+         makes the module type of %a depend on@ %t.@ \
+         Such recursive definitions of module types are not allowed.@]"
```

Comment:
I would be personally in favor of completely splitting the self-recursion case from the generic case, since this lead to more fluid message:
```ocaml
       "@[<hov>This module type is recursive.@ \
         This use of the recursive module %a@ \
         within its own definition@ \
         makes the module type of %a depend on itself.@ \
         Such recursive definitions of module types are not allowed.@]"
         ```
         But the current state is fine with me.